### PR TITLE
Fix false num_classes warning in metrics

### LINKED ISSUE #2768: to_categorical should go before get_num_classes in metrics/functional/classification.py
https://github.com/PyTorchLightning/pytorch-lightning/blob/d18b9ef9d95ffd92591f767808cc497af5bd4e1c/pytorch_lightning/metrics/functional/classification.py#L174-L178

`to_categorical` should go before `get_num_classes`, since `get_num_classes` assumes `pred` has already been turned into categories (it is doing `int(pred.max().detach().item() + 1)`).

Warnings are raised now and then with current code, for example:
```
UserWarning: You have set 10 number of classes if different from predicted (xx) and target (10) number of classes
```

If logits are passed in directly for metrics.

### PR DESCRIPTION
**Pull request by pwwang**
## What does this PR do?

Put to_categorical before get_num_classes in metrics/functional/classification.py, which fixes false warning about num_classes.

Fixes #2768

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] (no need) Did you make sure to update the documentation with your changes?
- [ ] (no need) Did you write any new necessary tests? 
- [ ] (no need) Did you verify new and existing tests pass locally with your changes?
- [ ] (no need) If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by edenlightning** [comment]
Can you fix failing CIs?

**Comment by Borda** [comment]
> Can you fix failing CIs?

@ananyahjha93 is rebasing it, these issues have been fixed today on master :]

**Comment by Borda** [comment]
> # [Codecov](https://codecov.io/gh/PyTorchLightning/pytorch-lightning/pull/2781?src=pr&el=h1) Report
> > Merging [#2781](https://codecov.io/gh/PyTorchLightning/pytorch-lightning/pull/2781?src=pr&el=desc) into [master](https://codecov.io/gh/PyTorchLightning/pytorch-lightning/commit/3772601cd6872cde006aab9284e103e857955457&el=desc) will **decrease** coverage by `4%`.
> > The diff coverage is `100%`.
> 
> ```diff
> @@           Coverage Diff           @@
> ##           master   #2781    +/-   ##
> =======================================
> - Coverage      91%     86%    -4%     
> =======================================
>   Files          76      76            
>   Lines        6787    6787            
> =======================================
> - Hits         6150    5863   -287     
> - Misses        637     924   +287     
> ```

this is strange, it speaks the stats even the condition of waiting for all 22 builds is not fulfill, just checked codecov and it has only 18 builds and still, it speaks...
cc: @awaelchli @williamFalcon
